### ISSUE TITLE
How to configure RowFileInputReader to detect new files when existing file has not been timed out (read.max.wait.ms)?

### ISSUE THREAD

**Issue by shinlim** [streamthoughts/kafka-connect-file-pulse#83]
I have been experimenting with various configurations to implement tailing log files which rollover hourly e.g.

```
2020-10-26.0.log
2020-10-26.1.log
...
2020-10-26.23.log
```

There can be periods of inactivity which mean a log file isn't updated for several tens of minutes. If an event is logged after the hour, it will be logged in a new file.

When I used the below configuration, it resulted in the file connector waiting for an hour on the previous file, and not ingesting the new file:

```
{
  "name": "event-log-source",
  "config": {
    "connector.class": "io.streamthoughts.kafka.connect.filepulse.source.FilePulseSourceConnector",
    "fs.cleanup.policy.class": "io.streamthoughts.kafka.connect.filepulse.clean.DeleteCleanupPolicy",
    "fs.scanner.class": "io.streamthoughts.kafka.connect.filepulse.scanner.local.LocalFSDirectoryWalker",
    "fs.scan.directory.path": "/var/log/events",
    "fs.scan.interval.ms": "5000",
    "fs.scan.filters": "io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.RegexFileListFilter,io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.LastModifiedFileListFilter",
    "fs.recursive.scan.enable": "false",
    "file.filter.regex.pattern":  ".*\\.log$",
    "file.filter.minimum.age.ms": "10000",
    "internal.kafka.reporter.bootstrap.servers": "localhost:9092",
    "internal.kafka.reporter.topic": "event-sender-status",
    "offset.strategy": "path+name+inode",
    "read.max.wait.ms": "3600000",
```

It seems the connector only watches 1 file at a time, even if I use multiple tasks configured as 'tasks.max: 5'

Is there a way to achieve this behaviour using the same connector configuration? I have also tried the below config, without success:

```
{
  "name": "event-log-source",
  "config": {
    "connector.class": "io.streamthoughts.kafka.connect.filepulse.source.FilePulseSourceConnector",
    "fs.cleanup.policy.class": "io.streamthoughts.kafka.connect.filepulse.clean.LogCleanupPolicy",
    "fs.scanner.class": "io.streamthoughts.kafka.connect.filepulse.scanner.local.LocalFSDirectoryWalker",
    "fs.scan.directory.path": "/var/log/events",
    "fs.scan.interval.ms": "5000",
    "fs.scan.filters": "io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.RegexFileListFilter,io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.LastModifiedFileListFilter",
    "fs.recursive.scan.enable": "false",
    "file.filter.regex.pattern":  ".*\\.log$",
    "file.filter.minimum.age.ms": "10000",
    "internal.kafka.reporter.bootstrap.servers": "localhost:9092",
    "internal.kafka.reporter.topic": "event-sender-status",
    "offset.strategy": "path+name+inode",
    "read.max.wait.ms": "60000",
...
{
  "name": "event-log-source-cleaner",
  "config": {
    "connector.class": "io.streamthoughts.kafka.connect.filepulse.source.FilePulseSourceConnector",
    "fs.cleanup.policy.class": "io.streamthoughts.kafka.connect.filepulse.clean.DeleteCleanupPolicy",
    "fs.scanner.class": "io.streamthoughts.kafka.connect.filepulse.scanner.local.LocalFSDirectoryWalker",
    "fs.scan.directory.path": "/var/log/events",
    "fs.scan.interval.ms": "5000",
    "fs.scan.filters": "io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.RegexFileListFilter,io.streamthoughts.kafka.connect.filepulse.scanner.local.filter.LastModifiedFileListFilter",
    "fs.recursive.scan.enable": "false",
    "file.filter.regex.pattern":  ".*\\.log$",
    "file.filter.minimum.age.ms": "10000",
    "internal.kafka.reporter.bootstrap.servers": "localhost:9092",
    "internal.kafka.reporter.topic": "event-sender-status",
    "offset.strategy": "path+name+inode",
    "read.max.wait.ms": "3600000",
```

The idea with the above config is to have a 'quicker' connector scan the directory for new files and the 'slower' connector wait until the previous hour's data had been committed to Kafka before deleting the file which is no longer needed when the data has been pushed to Kafka. Since the connectors share state in the Kafka status topics, they should cooperate over the same log file and move onto new files when they become available - but I may have misunderstood the documentation.

Can you advise me please on how I can achieve the behaviour above?

Thanks in advance!

Best regards
Shin

**Comment by fhussonnois** [comment]
Hi @shinlim, well, I think you've hit on one of the current limitations of Connect FilePulse. Currently the connector will not schedule new tasks until all previous taks (i.e files) has been completed even if new files has been detected.

It should be relatively easy to update the connector to add (for example) this new parameter: 

- `allow.tasks.schedule.after.timeout.ms` : Specifies the timeout (in miliseconds) for the connector to allow tasks re-scheduling when new files are detected, and some tasks are still being processing. 

What do you think?

**Comment by shinlim** [comment]
Thank you! I will try this out and post back the result.

**Comment by shinlim** [comment]
That would be a very useful feature! Would we be able to configure that on only 1 connector so that if we specified multiple tasks, would that mean while 1 task was waiting on the file to complete, others could pick up new files?

Thank you for your support!

**Comment by fhussonnois** [comment]
Yes that would be the idea. So for example with `allow.tasks.schedule.after.timeout.ms=0` then if a new file is detected all tasks will be re-configured immediately.

I will work on that new feature as soon as possible this week.

**Comment by fhussonnois** [comment]
Hi @shinlim, I'm closing this issue because the new parameter is available in this new release : https://github.com/streamthoughts/kafka-connect-file-pulse/releases/tag/v1.6.0

The new property is `allow.tasks.reconfiguration.after.timeout.ms`

**Comment by shinlim** [comment]
Thank you so much for developing this feature! I will test it out and provide feedback in the next week.

Best regards
Shin
